src/systems/userinput/userinput.js

```diff
@@ -0,0 +1,188 @@
+      if (AFRAME.utils.device.isMobile()) {
+        if (inVRMode) {
+          this.activeDevices.delete(appAwareTouchscreenDevice);
+          this.registeredMappings.delete(touchscreenUserBindings);
+        } else {
+          this.activeDevices.add(appAwareTouchscreenDevice);
+          this.registeredMappings.add(touchscreenUserBindings);
+        }
+      } else {
+        if (inVRMode) {
+          this.registeredMappings.delete(keyboardMouseUserBindings);
+        } else {
+          this.registeredMappings.add(keyboardMouseUserBindings);
+        }
+      }
+    };
+    this.el.sceneEl.addEventListener("enter-vr", updateBindingsForVRMode);
+    this.el.sceneEl.addEventListener("exit-vr", updateBindingsForVRMode);
+    updateBindingsForVRMode();
+
+    window.addEventListener(
+      "gamepadconnected",
+      e => {
+        console.log(e.gamepad);
+        let gamepadDevice;
+        if (e.gamepad.id === "OpenVR Gamepad") {
+          for (let i = 0; i < this.activeDevices.length; i++) {
+            const activeDevice = this.activeDevices[i];
+            if (activeDevice.gamepad && activeDevice.gamepad === e.gamepad) {
+              console.warn("ignoring gamepad");
+              return; // multiple connect events without a disconnect event
+            }
+          }
+          if (this.activeDevices) gamepadDevice = new ViveControllerDevice(e.gamepad);
+          this.registeredMappings.add(viveUserBindings);
+        } else if (e.gamepad.id.startsWith("Oculus Touch")) {
+          gamepadDevice = new OculusTouchControllerDevice(e.gamepad);
+          this.registeredMappings.add(oculusTouchUserBindings);
+        } else if (e.gamepad.id === "Oculus Go Controller") {
+          gamepadDevice = new OculusGoControllerDevice(e.gamepad);
+          this.registeredMappings.add(oculusGoUserBindings);
+        } else if (e.gamepad.id === "Daydream Controller") {
+          gamepadDevice = new DaydreamControllerDevice(e.gamepad);
+          this.registeredMappings.add(daydreamUserBindings);
+        } else if (e.gamepad.id.includes("Xbox")) {
+          gamepadDevice = new XboxControllerDevice(e.gamepad);
+          this.registeredMappings.add(xboxControllerUserBindings);
+        } else {
+          gamepadDevice = new GamepadDevice(e.gamepad);
+          this.registeredMappings.add(gamepadBindings);
+        }
+        this.activeDevices.add(gamepadDevice);
+        this.gamepads[e.gamepad.index] = gamepadDevice;
+      },
+      false
+    );
+    window.addEventListener(
+      "gamepaddisconnected",
+      e => {
+        if (this.gamepads[e.gamepad.index]) {
+          this.activeDevices.delete(this.gamepads[e.gamepad.index]);
+          delete this.gamepads[e.gamepad.index];
+        }
+      },
+      false
+    );
+  },
+
+  tick() {
+    updateActionSetsBasedOnSuperhands();
+
+    for (const { set, value } of this.pendingSetChanges) {
+      this.activeSets[value ? "add" : "delete"](set);
```

Comment:
Yeah, originally we had set changes in more spots, they all converged to this one spot. Unclear if this will always be the case or if it will be nice to be able to queue up actionset changes from arbitrary spots in code.